balancer.go

```diff
@@ -40,19 +41,61 @@ func (n *Notification) claim(current map[string][]int32) {
 
 // --------------------------------------------------------------------
 
-type topicInfo struct {
+var (
+	balancerStrategies map[Strategy]Balancer = map[Strategy]Balancer{
+		StrategyRange:      &RangeBalancer{},
+		StrategyRoundRobin: &RoundRobinBalancer{},
+		StrategyStriped:    &StripedBalancer{},
+	}
+)
+
+// RegisterBalancerStrategy allows for use of custom rebalance strategies.
+func RegisterBalancerStrategy(name Strategy, b Balancer) {
+	balancerStrategies[name] = b
+}
+
+// --------------------------------------------------------------------
+
+// Balancer implements a method Rebalance that determines which topics/partitions each
+// member of the group should consume.
+type Balancer interface {
+	// Rebalance takes a map of topic names with group member and partition information, and
+	// returns a map of group members with a map of the topics/partitions they should consume.
+	// Example input:
+	// Topic1:
+	//   MemberIDs: [Consumer1, Consumer2]
+	//   Partitions: [0, 1]
+	// Topic2:
+	//   MemberIDs: [Consumer2, Consumer3]
+	//   Partitions: [0, 1, 2, 3]
+	// Example output (in this case, StrategyStriped):
+	// Consumer1:
+	//   Topic1: [0]
+	//   Topic2: [1]
+	// Consumer2:
+	//   Topic1: [1]
+	//   Topic2: [2]
+	// Consumer3:
+	//   Topic2: [0, 3]
+	Rebalance(topics map[string]TopicInfo) map[string]map[string][]int32
+}
+
+// TopicInfo contains the partitions a topic contains and the member IDs subscribed to it.
+type TopicInfo struct {
+	Topic      string
 	Partitions []int32
 	MemberIDs  []string
 }
 
-func (info topicInfo) Perform(s Strategy) map[string][]int32 {
-	if s == StrategyRoundRobin {
-		return info.RoundRobin()
-	}
-	return info.Ranges()
+// RangeBalancer implements StrategyRange, balancing partitions in ranges to consumers.
+type RangeBalancer struct{}
```

Comment:
I left these as public because one could incorporate them into their own balancer if needed (e.g., if you had topics A, B, C, you could use a unique strategy for topic A while letting B and C use a built-in strategy).